rfcs/2024-08-21-motion-presets.md

```diff
@@ -0,0 +1,523 @@
+};
+```
+
+#### Move
+
+```jsx
+import { Move } from '@razorpay/blade/components';
+
+<Move isVisible={showCard}>
+  <Card>
+    <CardBody>
+      <Text>Fade In/Out Card</Text>
+    </CardBody>
+  </Card>
+</Move>;
+```
+
+```ts
+type MoveProps = {
+  /**
+   * @default inout
+   */
+  variant: 'in' | 'out' | 'inout';
+  /**
+   * @default 'bottom'
+   */
+  direction: 'top' | 'right' | 'bottom' | 'left';
+
+  /**
+   * Visibility state
+   */
+  isVisible?: boolean;
+};
+```
+
+</details>
+
+### Highlight Animations
+
+#### Morph
+
+> [!Note]
+>
+> Defining morph as preset is a bit tricky and need to make sure the API is feasible with all possible scenarios. Currently we've done a basic feasibility check but we might change / update the API if we come across some animation that can't be built with this API.
+
+<table>
+<tr>
+  <td width="600px">
+In below example, the images scales up when its parent container is hovered<br/><br/>
+
+<!-- prettier-ignore -->
+```jsx
+import { AnimatePresence } from 'framer-motion';
+import { 
+  Button, 
+  Morph 
+} from '@razorpay/blade/components';
+
+<AnimatePresence>
+{
+  isChatVisible ? (
+    <Morph key="chat" layoutId="chat-interface">
+      <ChatInterface />
+    </Morph>
+  ) : (
+    <Morph key="chat-button" layoutId="chat-interface">
+      <Button icon={RazorpayIcon} />
+    </Morph>
+  )
+}
+</AnimatePresence>
+```
+
+  </td>
+  <td width="500px">
+
+https://github.com/user-attachments/assets/2ec967a5-1e17-443a-a3a9-230962f6bd9a
+
+_Previews are just examples of presets. They don't use actual durations and easings yet_
+
+  </td>
+</tr>
+</table>
+
+<details>
+<summary>Alternate Morph APIs</summary>
+
+#### 2. Using `motion` from framer-motion
+
+```jsx
+import { motion } from 'framer-motion';
+import { Heading } from '@razorpay/blade/components';
+
+const CardHeading = motion(Heading);
+
+<CardHeading layoutId="card-heading" transition={{ duration: theme.motion.duration.slow }}>Hello, World!</CardHeading>
+<CardHeading as="h1" layoutId="card-heading">Hello, World!</CardHeading>
+```
+
+**Cons:**
+
+- We won't be able to preset the animation styles and durations with this approach
+- Can lead to inconsistent animations
+- Inconsistent with other motion presets so not very intuitive and requires learning framer-motion for syntax
+
+#### 3. Exposing `morph` Function from Blade
+
+```jsx
+import { Heading, morph } from '@razorpay/blade/components';
+
+const CardHeading = morph(Heading);
+
+<CardHeading layoutId="card-heading">Hello, World!</CardHeading>
+<Card>
+  <CardBody>
+    <CardHeading as="h1" layoutId="card-heading">Hello, World!</CardHeading>
+    <Box>
+      <Text>Other Text</Text>
+    </Box>
+  </CardBody>
+</Card>
+```
+
+**Pros:**
+
+- Allows presetting animation properties on blade
+
+**Cons:**
+
+- Inconsistent with other motion presets so not very intuitive
+- Comparitively more verbose than suggested API
+
+</details>
+
+#### AnimateInteractions
```

Comment:
What does AnimateInteractions do internally? Can you provide an overview?

Reply:
It creates motion.div with whileHover, whileTap, whileFocus etc (depending on which interactions you want. Taking it as a prop), it defines the name of the variant of its child.

You can check this POC - https://codesandbox.io/p/sandbox/framer-motion-hover-presets-poc-ghlcpl

The Card component in this example is basically the AnimateInteractions